plugins/org.eclipse.n4js.flowgraphs/src/org/eclipse/n4js/flowgraphs/factories/ControlFlowGraphFactory.java

```diff
@@ -86,8 +99,19 @@ static private void createComplexNodes(Script script, TreeSet<ControlFlowElement
 
 	static private void connectComplexNodes(ComplexNodeMapper cnMapper) {
 		for (ComplexNode cn : cnMapper.getAll()) {
-			for (Node mNode : cn.getAllButExitNodes()) {
-				connectNode(cnMapper, mNode);
+			List<Node> removeNodes = new LinkedList<>();
+			for (Node mNode : cn.getNodes()) {
+				if (mNode != cn.getExit()) {
+					connectNode(cnMapper, mNode);
+				}
+			}
+			for (Node mNode : cn.getNodes()) {
+				if (removeDelegateNode(cn, mNode)) {
+					removeNodes.add(mNode);
```

Comment:
do you need this intermediate collection at all?

Reply:
yes. as discussed